### PR TITLE
Highlight aliases once

### PR DESCRIPTION
**Pull request by phy1729**

### CODE REVIEW

File: highlighters/main/test-data/alias-precommand-option-argument1.zsh [hunk 1]
```diff
@@ -35,7 +35,6 @@ BUFFER='sdu phy1729 echo foo'
 
 expected_region_highlight=(
   '1 3 alias' # sdu
-  '1 3 precommand' # sdu (sudo)
   '5 11 default' # phy1729
   '13 16 commmand "issue #540"' # echo (not builtin)
```
Comment (danielshahaf) [inline highlighters/main/test-data/alias-precommand-option-argument1.zsh]: Note that if `echo` ceased being highlighted as `builtin`, the test wouldn't fail. Can we use, say, `ls` here to test this situation, and add a separate XFail test for the second word of `sudo echo` (without aliases)?

File: highlighters/main/test-data/alias-precommand-option-argument2.zsh [hunk 2]
```diff
@@ -28,15 +28,14 @@
 # vim: ft=zsh sw=2 ts=2 et
 # -------------------------------------------------------------------------------------------------
 
-alias sde='sudo -e'
-alias seu='sde -u'
+alias sdb='sudo -b'
+alias sbu='sdb -u'
```
Comment (danielshahaf) [inline highlighters/main/test-data/alias-precommand-option-argument2.zsh]: Those cryptic names are a bit hard to follow. Not a big issue but could you use longer names in the future?

File: highlighters/main/test-data/alias-precommand-option-argument2.zsh [hunk 3]
```diff
@@ -28,15 +28,14 @@
...
 
-BUFFER='seu phy1729 echo foo'
+BUFFER='sbu phy1729 echo foo'
 
 expected_region_highlight=(
-  '1 3 alias' # seu
-  '1 3 precommand' # seu (sudo)
+  '1 3 alias' # sbu
   '5 11 default' # phy1729
   '13 16 commmand "issue #540"' # echo (not builtin)
```
Comment (danielshahaf) [inline highlighters/main/test-data/alias-precommand-option-argument2.zsh]: Ditto (use `ls`).

File: highlighters/main/test-data/alias-to-dir.zsh [hunk 4]
```diff
@@ -33,5 +33,4 @@ BUFFER=$'x'
 
 expected_region_highlight=(
   '1 1 alias' # x
-  '1 1 unknown-token "issue #202"' # x (/)
```
Comment (danielshahaf) [inline highlighters/main/test-data/alias-to-dir.zsh]: Should expect unknown-token here?

File: highlighters/main/test-data/empty-command.zsh [hunk 5]
```diff
@@ -32,5 +32,5 @@ BUFFER='echo; ;'
 expected_region_highlight=(
   "1 4 builtin" # echo
   "5 5 commandseparator" # ;
-  "7 7 unknown-token" # ;
+  "7 7 commandseparator" # ;
```
Comment (danielshahaf) [inline highlighters/main/test-data/empty-command.zsh]: Why? Doesn't seem related to aliases.

File: highlighters/main/test-data/empty-command2.zsh [hunk 6]
```diff
@@ -35,5 +35,5 @@ BUFFER='echo; ;'
 expected_region_highlight=(
   "1 4 builtin" # echo
   "5 5 commandseparator" # ;
-  "7 7 unknown-token" # ;
+  "7 7 commandseparator" # ;
```
Comment (danielshahaf) [inline highlighters/main/test-data/empty-command2.zsh]: Ditto

File: highlighters/main/test-data/empty-line.zsh [hunk 7]
```diff
@@ -31,6 +31,6 @@
 BUFFER=$'\\\n; ls'
 
 expected_region_highlight=(
-  '3 3 unknown-token' # ;
+  '3 3 commandseparator' # ;
```
Comment (danielshahaf) [inline highlighters/main/test-data/empty-line.zsh]: Unrelated: I wonder if we should add a style for the backslash-newline.

File: highlighters/main/main-highlighter.zsh [hunk 8]
```diff
@@ -73,8 +73,10 @@ _zsh_highlight_main_add_region_highlight() {
   integer start=$1 end=$2
   shift 2
 
-  (( highlighted_alias )) && return
-  (( in_alias )) && highlighted_alias=1
+  if (( in_alias )); then
+    [[ $1 == unknown-token ]] && alias_style=unknown-token
```
Comment (danielshahaf) [inline highlighters/main/main-highlighter.zsh]: Is this tested?

For example

```
alias a=b b=foo
BUFFER='a '
expected_region_highlight=(
  '1 1 unknown-token'
)
```

```
alias a='() { ls "$@" ; foo }'
BUFFER='a '
expected_region_highlight=(
  '1 1 unknown-token'
)
```

File: highlighters/main/test-data/empty-command.zsh (hunk 5 above)
Comment (danielshahaf) [reply highlighters/main/test-data/empty-command.zsh]: See also #616

File: highlighters/main/test-data/alias-precommand-option-argument1.zsh (hunk 1 above)
Comment (phy1729) [reply highlighters/main/test-data/alias-precommand-option-argument1.zsh]: Added alias-precommand-option-argument3.

File: highlighters/main/test-data/alias-precommand-option-argument2.zsh (hunk 2 above)
Comment (phy1729) [reply highlighters/main/test-data/alias-precommand-option-argument2.zsh]: Fixed.

File: highlighters/main/test-data/empty-command.zsh (hunk 5 above)
Comment (phy1729) [reply highlighters/main/test-data/empty-command.zsh]: No longer changed.

File: highlighters/main/test-data/empty-command2.zsh (hunk 6 above)
Comment (phy1729) [reply highlighters/main/test-data/empty-command2.zsh]: No longer changed.

File: highlighters/main/test-data/empty-line.zsh (hunk 7 above)
Comment (phy1729) [reply highlighters/main/test-data/empty-line.zsh]: I wouldn't be opposed, though it'd probably prompt allowing highlighting of all backslashes as well.

File: highlighters/main/main-highlighter.zsh (hunk 8 above)
Comment (phy1729) [reply highlighters/main/main-highlighter.zsh]: Added.

File: highlighters/main/test-data/alias-to-dir.zsh (hunk 4 above)
Comment (phy1729) [reply highlighters/main/test-data/alias-to-dir.zsh]: Yes, fixed.

**Comment by danielshahaf** [comment]
This fixes #576, which is a release blocker.

The log messages don't mention #576.

Shall we merge this?

**Comment by phy1729** [comment]
I'll fixup the log message and merge (and fix the commmand misspelling I noticed).

**Comment by danielshahaf** [comment]
Thanks!
